chips/sam4l/src/flashcalw.rs

```diff
@@ -946,6 +947,102 @@ impl hil::flash::Flash for FLASHCALW {
     }
 }
 
+impl hil::flash::FlashInfo for FlashCalw {
+    fn flash_size(&self) -> u32 {
+        self.get_flash_size()
+    }
+    fn num_pages(&self) -> u32 {
+        self.get_page_count()
+    }
+    fn page_size(&self) -> u32 {
+        self.get_page_size()
+    }
+    fn num_lock_units(&self) -> u32 {
+        NB_OF_REGIONS
+    }
+    fn lock_unit_size(&self) -> u32 {
+        self.flash_size() / self.num_lock_units()
+    }
+    fn pages_per_lock_unit(&self) -> u32 {
+        self.get_page_count_per_region()
+    }
+    fn page_to_lock_unit(&self, page: u32) -> u32 {
+        self.get_page_region(page as i32)
+    }
+}
+
+impl hil::flash::FlashLocking for FlashCalw {
+    fn lock_unit(&self, unit: u32) {
+        self.lock_region(unit as u32, true);
+    }
+    fn unlock_unit(&self, unit: u32) {
+        self.lock_region(unit as u32, false);
+    }
+
+    /// Locks [first,last]
+    fn lock_units(&self, first: u32, last: u32) {
+        for unit in first..last + 1 {
+            self.lock_unit(unit)
+        }
+    }
+    fn unlock_units(&self, first: u32, last: u32) {
+        for unit in first..last + 1 {
+            self.lock_unit(unit)
+        }
+    }
+}
+
+impl hil::flash::FlashLayout for FlashCalw {
+    fn kernel_start_address(&self) -> u32 {
+        unsafe {
+            extern "C" {
+                static _stext: *const u32;
+            }
+            (&_stext as *const *const u32) as u32
+        }
+    }
+
+    fn kernel_end_address(&self) -> u32 {
+        unsafe {
+            extern "C" {
+                static _etext: *const u32;
+            }
+            (&_etext as *const *const u32) as u32
+        }
+    }
+
+    fn kernel_first_page(&self) -> u32 {
+        self.kernel_start_address() / self.get_page_size()
+    }
+
+    fn kernel_last_page(&self) -> u32 {
+        self.kernel_end_address() / self.get_page_size()
+    }
+
+    fn apps_start_address(&self) -> u32 {
+        unsafe {
+            extern "C" {
+                static _sapps: *const u32;
+            }
+            (&_sapps as *const *const u32) as u32
+        }
+    }
+
+    fn apps_end_address(&self) -> u32 {
+        self.get_flash_size() - 1
```

Comment:
If this isn't correct then it should be reflected in the linker map; we should add a symbol to kernel_layout.ld that reflects this. It shouldn't be a magic constant in software.